src/rtcModule/webrtcAdapter.cpp

```diff
@@ -78,171 +88,382 @@ unsigned long generateId()
-        devices.push_back(cricket::Device("default", "0"));
+        char deviceName[256];
+        char uniqueName[256];
+        std::unique_ptr<webrtc::VideoCaptureModule::DeviceInfo> deviceInfo(webrtc::VideoCaptureFactory::CreateDeviceInfo());
+        deviceInfo->GetDeviceName(i, deviceName, sizeof(deviceName), uniqueName, sizeof(uniqueName));
+        videoDevices.insert(std::pair<std::string, std::string>(deviceName, uniqueName));
     }
+
+    return videoDevices;
 }
 
-template<>
-void InputDeviceShared<webrtc::VideoTrackInterface, webrtc::VideoTrackSourceInterface>::createSource()
+void CaptureModuleLinux::openDevice(const std::string &videoDevice)
 {
-    assert(!mSource.get());
-
-#ifndef __ANDROID__
-    cricket::WebRtcVideoDeviceCapturerFactory factory;
-    std::unique_ptr<cricket::VideoCapturer> capturer(
-        factory.Create(cricket::Device(mOptions->device.name, 0)));
-    if (!capturer)
+    mCameraCapturer = webrtc::VideoCaptureFactory::Create(videoDevice.c_str());
+    if (!mCameraCapturer)
     {
-        RTCM_LOG_WARNING("Could not create video capturer for device '%s'", mOptions->device.name.c_str());
         return;
     }
 
-    mCapturer = capturer.get();
-    mSource = gWebrtcContext->CreateVideoSource(capturer.release(),
-        &(mOptions->constraints));
-#else
-    JNIEnv *env;
-    MEGAjvm->AttachCurrentThread(&env, NULL);
-    rtc::Thread *currentThread = rtc::ThreadManager::Instance()->CurrentThread();
-    mSource = new rtc::RefCountedObject<webrtc::AndroidVideoTrackSource>(currentThread, env, surfaceTextureHelper, false);
-    rtc::scoped_refptr<webrtc::VideoTrackSourceProxy> proxySource = webrtc::VideoTrackSourceProxy::Create(currentThread, currentThread, mSource);
-    if (startVideoCaptureWithParametersMID)
+    mCameraCapturer->RegisterCaptureDataCallback(this);
+
+    std::unique_ptr<webrtc::VideoCaptureModule::DeviceInfo> deviceInfo(webrtc::VideoCaptureFactory::CreateDeviceInfo());
+    webrtc::VideoCaptureCapability capabilities;
+    deviceInfo->GetCapability(mCameraCapturer->CurrentDeviceName(), 0, capabilities);
+    mCapabilities.interlaced = capabilities.interlaced;
+    mCapabilities.videoType = webrtc::VideoType::kI420;
+
+    if (mCameraCapturer->StartCapture(mCapabilities) != 0)
     {
-        std::string widthString;
-        mOptions->constraints.GetMandatory().FindFirst(webrtc::MediaConstraintsInterface::kMinWidth, &widthString);
-        int width = std::atoi(widthString.c_str());
-        std::string heightString;
-        mOptions->constraints.GetMandatory().FindFirst(webrtc::MediaConstraintsInterface::kMinHeight, &heightString);
-        int height = std::atoi(heightString.c_str());
-        int fps = 15;
-        env->CallStaticVoidMethod(applicationClass, startVideoCaptureWithParametersMID, (jint)width, (jint)height, (jint)fps, (jlong)proxySource.release(), surfaceTextureHelper);
+        return;
```

Comment:
We should add an error message to the log here

Reply:
Fixed